internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,221 @@
+		Name:      installEnv.Spec.ClusterRef.Name,
+		Namespace: installEnv.Spec.ClusterRef.Namespace,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+
+	if err := r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		clusterDeploymentRefErr := newKubeAPIError(
+			errors.Wrapf(
+				err,
+				fmt.Sprintf("failed to find clusterDeployment with name %s in namespace %s",
+					installEnv.Spec.ClusterRef.Name, installEnv.Spec.ClusterRef.Namespace)),
+			k8serrors.IsNotFound(err),
+		)
+		return r.updateStatusAndReturnResult(ctx, installEnv, nil, imageState, clusterDeploymentRefErr)
+	}
+
+	c, err := r.Installer.GetClusterByKubeKey(types.NamespacedName{
+		Name:      clusterDeployment.Name,
+		Namespace: clusterDeployment.Namespace,
+	})
+
+	if gorm.IsRecordNotFoundError(err) {
+		inventoryErr = common.NewApiError(http.StatusNotFound, err)
+	} else if err != nil {
+		inventoryErr = common.NewApiError(http.StatusInternalServerError, err)
+	} else {
+		isoParams := installer.GenerateClusterISOParams{
+			ClusterID:         *c.ID,
+			ImageCreateParams: &models.ImageCreateParams{},
+		}
+		if len(installEnv.Spec.SSHAuthorizedKeys) > 0 {
+			isoParams.ImageCreateParams.SSHPublicKey = installEnv.Spec.SSHAuthorizedKeys[0]
+		}
+		// GenerateClusterISOInternal will generate an ISO only if there it was not generated before,
+		// or something has changed in isoParams.
+		updatedCluster, inventoryErr = r.Installer.GenerateClusterISOInternal(ctx, isoParams)
+	}
+
+	if inventoryErr == nil {
+		imageInfo = updatedCluster.ImageInfo
+		imageState.Status = corev1.ConditionTrue
+		imageState.State = ImageCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, installEnv, imageInfo, imageState, inventoryErr)
+}
+
+func (r *InstallEnvReconciler) updateStatusAndReturnResult(
+	ctx context.Context,
+	installEnv *adiiov1alpha1.InstallEnv,
+	imageInfo *models.ImageInfo,
+	imageState ImageState,
+	err error) (ctrl.Result, error) {
+
+	var res ctrl.Result
+
+	if imageBeingCreated(err) {
+		imageState.Status = corev1.ConditionTrue
+		imageState.State = ImageCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+		// Clear up the error stateInfo while installEnv is being created
+		err = nil
+		r.Log.Infof("Image %s being prepared for cluster %s stateInfo: %s",
+			installEnv.Name, installEnv.ClusterName, imageState.StateInfo)
+	} else if isClientError(err) {
+		imageState.Status = corev1.ConditionFalse
+		imageState.State = ImageCreationError
+		imageState.StateInfo += ": " + err.Error()
+	} else if err != nil {
+		imageState.Status = corev1.ConditionFalse
+		imageState.State = ImageCreationError
+		imageState.StateInfo += ": internal error"
+		res.Requeue = true
+	}
+	r.setStateAndStateInfo(adiiov1alpha1.ImageCreatedCondition, imageState, installEnv)
+
+	if err == nil && imageInfo != nil {
+		installEnv.Status.ISODownloadURL = imageInfo.DownloadURL
+	} else if err != nil {
```

Comment:
in case or error you probably need to remove the existing URL no?

Reply:
Yup